exonum/src/api/mod.rs

```diff
@@ -295,3 +278,46 @@ pub trait Api {
     /// Used to extend Api.
     fn wire<'b>(&self, router: &'b mut Router);
 }
+
+fn url_fragment<T>(request: &Request, name: &str) -> Result<T, ApiError>
+where
+    T: FromStr,
+    T::Err: ::std::error::Error + Send + Sync + 'static,
+{
+    let params = request.extensions.get::<Router>().unwrap();
+    let fragment = params.find(name).ok_or_else(|| {
+        ApiError::BadRequest(format!("Required parameter '{}' is missing", name))
+    })?;
+    let value: T = T::from_str(fragment).map_err(|e| {
+        ApiError::BadRequest(format!("Invalid '{}' parameter: {}", name, e))
```

Comment:
Could it make sense to create separate `ApiError` variants related to parameter processing? IMO, it could be clearer.

Reply:
Moderately strongly disagree here :-)

The sole purpose of `ApiError` is to be converted to `IronError`, whose sole purpose is to be converted to non-200/300 HTTP response. So the domain object that `ApiError` is modeling is actually an unsuccessful HTTP response, and there's no  specific status code for "invalid parameter".

I am not sure that `ApiError` enum is actually needed at all: it is just a disguised DSL to make working with hyper more pleasant. This enum has a ton of variants, but we never match it. Each ApiError instance created is almost instantly turned into IronError via a From impl and `?`.

For the same reason, I am not sure that exposing `url_fragment` and `requred/optional_parameter` is a good idea: this helpers have nothing to do with Exonum and are strictly iron specific. 

But perhaps practicality beats purity and we should make them pub with a comment like "yeah, this is a bad API which has nothing to do with Exonum, and it should not be considered stable, but you may use it, if you find it useful"?